### PR TITLE
CRM-20430 save report criteria

### PR DESCRIPTION
**Pull request by johanv**
This PR introduces a new permission for CiviReport: 'save Report Criteria'.
(see also CRM-20430).

---

 * [CRM-20430: A permission 'save Report Criteria'](https://issues.civicrm.org/jira/browse/CRM-20430)

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
I think this is ok from a code level - if someone is up for doing some UI testing of it

**Comment by seamuslee001** [comment]
Agree it looks good from a code perspective, I believe that we should like when we previously added a new Permission add a message as part of CRM_Upgrade_Incremental_php_FourSeven::setPostUpgradeMessage I think is the function. 

That way a message is clearly displayed to people so they know they have to go looking to perhaps add the new permission in

**Comment by eileenmcnaughton** [comment]
Ok so this is blocked on adding an upgrade message -

**Comment by mlutfy** [comment]
@johanv Any luck you  could add the missing upgrade message? Seems like this PR is otherwise ready to merge.

**Comment by johanv** [comment]
This got out of my sight since I have a new job. I'm not sure when I can add this upgrade message, but feel free to add one yourself.

**Comment by eileenmcnaughton** [comment]
@seamuslee001 do you feel enough love to do the upgrade message - if so I can merge that PR - if this is still sound after nearly 1 year

**Comment by seamuslee001** [comment]
i'll work on this shortly

**Comment by johanv** [comment]
If the upgrade message is the only thing that's missing, I think I can find some time this week to add one. So if you want me to do this, @seamuslee001, let me know.
I was looking at the code, and it seems that previously this kind of message was shown [as pre upgrade message](https://github.com/civicrm/civicrm-core/blob/master/CRM/Upgrade/Incremental/php/FourSeven.php#L63). So maybe we should use a pre upgrade message again (instead of a post upgrade message, as suggested above).

**Comment by eileenmcnaughton** [comment]
@johanv yes - a pre-upgrade message like that looks perfect

**Comment by johanv** [comment]
I added the pre-upgrade message, and I rebased, to prevent merge conflicts with the pre-commit message.

**Comment by eileenmcnaughton** [comment]
Ohh the test fail is actually related. I think the issue is that the permission check needs to be 

if (!$params['check_permissions'] || !CRM_Core_Permission::check('save Report Criteria')) {

and we need to pass through that parameter from calling functions.

File: CRM/Report/BAO/ReportInstance.php [hunk 1]
```diff
@@ -41,10 +41,16 @@ class CRM_Report_BAO_ReportInstance extends CRM_Report_DAO_ReportInstance {
    *
    * @param array $params
    *   (reference ) an assoc array of name/value pairs.
-   *
-   * @return CRM_Report_DAO_ReportInstance
+   * @return \CRM_Report_DAO_ReportInstance
+   * @throws \CRM_Core_Exception
    */
   public static function add(&$params) {
+    if (!CRM_Core_Permission::check('save Report Criteria')) {
```
Comment (eileenmcnaughton) [inline CRM/Report/BAO/ReportInstance.php]: so to make the test pass I think this needs to be

if (CRM_Utils_Array::value('check_permissions') && !CRM_Core_Permission::check('save Report Criteria'))

Comment (eileenmcnaughton) [reply CRM/Report/BAO/ReportInstance.php]: Alternatively this check could be stripped out & we could reply on the form - which is consistent with a lot of other places ( you indicate that it doesn't really work anyway)

CRM_Report_Info::getPermissions is the place to define api permissions

**Comment by eileenmcnaughton** [comment]
Closing in favour of https://github.com/civicrm/civicrm-core/pull/12107

**Pull request closed without merging**